### ISSUE TITLE
Feature request: diff for Vector

### ISSUE THREAD

**Issue by ckaran** [bodil/im-rs#181]
It it possible to implement a difference or edit iterator for Vector?  I want to do something similar to 'rsync', but with the difference that each node in the system knows what the other node already knows.  This could be done pretty easily using Vector:

1. Create an original (empty) Vector, which will result in an empty patch set.  Send that to the 'other side'.
2. Clone the above Vector and mutate the clone.  Because Vectors are immutable, the clone is almost zero cost.
3. Mutate the clone as needed.
4. **The feature request ->** Iterate over the differences between the clone and the original periodically, creating an edit set.
5. Send the edit set to the 'other side'.
6. Once you receive confirmation that the 'other side' has updated itself to match your clone, discard the original Vector from step 1. 
7.  Rename the clone as the original.  
8. Goto step 2.

**Comment by ckaran** [comment]
I just ran across [diffus](https://crates.io/crates/diffus).  It may be of help here.

**Comment by ollpu** [comment]
This could also be very useful with maps.

In general, iterating over the data structure tree and looking for differences is not trivial, because nodes can also move, during tree rebalancing and such. I'm not too familiar with the implementations, but it might be impossible without storing more metadata for some of them.

**Comment by ckaran** [comment]
@ollpu Agreed!  Honestly, it would be great if there a trait that enabled this that all of the collections implemented.  E.g., something like:

```rust
pub trait Changes {
    /// Returns an iterator of changes between this version and the prior version.
    ///
    /// Returns an object that is able to iterate over the changes between this
    /// version and the immediate predecessor version.  The iterator object 
    /// **must** be valid even if the predecessor version or this version of the
    /// object are dropped.
    fn change_iterator(&self) -> Result<ChangeIterator, Error>;
}
```
